quantum/plugins/algorithms/qaoa/qaoa_maxcut.cpp

```diff
@@ -0,0 +1,323 @@
+                    positions[i] -= dTheta;
+                }
+            }
+            values.push_back(cutvalue);
+        }
+    }
+    std::pair<std::vector<double>, std::vector<double>> solution = std::make_pair(positions, values);
+    return solution;
+}
+
+std::pair<std::vector<double>, std::vector<double>> rotatesolution(std::pair<std::vector<double>, std::vector<double>> solution, int top)
+{   
+    std::vector<double> positions = std::get<0>(solution);
+    std::vector<double> values = std::get<1>(solution);
+    auto angles = regularize(positions[top-1], values[top-1]);
+    auto phi = std::get<0>(angles);
+    auto theta = std::get<1>(angles);
+    theta *= -1.;
+    std::vector<double> newPositions, newValues;
+    std::tuple<double,double,double> xyz;
+    std::pair<double,double> polar;
+    double newX, newY, newZ;
+    int i = 0;
+    for (auto position : positions)
+    {
+        position = std::fmod(position - phi, 2 * M_PI);
+        auto newvals = regularize(position, values[i]);
+        position = std::get<0>(newvals);
+        values[i] = std::get<1>(newvals);
+        xyz = polartocart(position, values[i]);
+        newX = std::cos(theta) * std::get<0>(xyz) * std::sin(theta) * std::get<2>(xyz);
+        newY = std::get<1>(xyz);
+        newZ = -1. * std::sin(theta) * std::get<0>(xyz) * std::cos(theta) * std::get<2>(xyz);
+        polar = carttopolar(newX,newY,newZ);
+        newPositions.push_back(std::get<0>(polar));
+        newValues.push_back(std::get<1>(polar));
+        i += 1;
+    }
+    phi = 2 * M_PI * std::rand();
+    for (int i=0; i < newPositions.size(); ++i)
+    {
+        newPositions[i] = std::fmod(newPositions[i] + phi, 2 * M_PI);
+    }
+    std::pair<std::vector<double>, std::vector<double>> rotated_solution = std::make_pair(newPositions, newValues);
+    return rotated_solution;
+} 
+// end warm_start functions
+
+namespace xacc {
+namespace algorithm {
+    // Construct the MAX-CUT Hamiltonian as an XACC observable
+    // based on input graphs edges
+    Observable *maxcut_qaoa::constructMaxCutHam(xacc::Graph *in_graph) const {
+        xacc::quantum::PauliOperator H;
+        for (int i = 0; i < in_graph->order(); ++i) {
+            auto neighbors = in_graph->getNeighborList(i);
+            for (const auto &neighborId : neighbors) {
+                if (in_graph->getEdgeWeight(i,neighborId)){
+                    auto weight = in_graph->getEdgeWeight(i,neighborId);
+                    H += 0.5 * weight * (xacc::quantum::PauliOperator(1.) - xacc::quantum::PauliOperator({{i, "Z"}, {neighborId, "Z"}}));
+                } else {
+                H += 0.5 * (xacc::quantum::PauliOperator(1.) - xacc::quantum::PauliOperator({{i, "Z"}, {neighborId, "Z"}})); 
+                }
+            }
+        }
+        // std::cout << "Hamiltonian: " << H.toString() << std::endl;
+        static auto graphHam = xacc::quantum::getObservable("pauli", H.toString());
+        return graphHam.get();
```

Comment:
Very minor: I think we should convert this function to return a `std::shared_ptr<Observable>` and let the `maxcut_qaoa` owns it rather than using `static`.
i.e. just return `xacc::quantum::getObservable("pauli", H.toString())`

The reason it was this way because in the original QAOA, the observable was normally created by the caller then passed to QAOA. I used static as a hacky way :) to support the Graph input.